restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceMethodDescriptor.java

```diff
@@ -430,6 +437,85 @@ public String getMethodName()
     return _type.toString();
   }
 
+  /**
+   * @return The type for the result computed by the method synchronously or asynchronously after removing any
+   * async wrappers like Task/Promise/Callback.
+   */
+  public Class<?> getMethodResultType() {
+    JavaMethodType methodType = getJavaMethodType();
+
+    switch (methodType) {
+      case ASYNC:
+        // Find the index of the callback param.
+        int callBackParamIndex = IntStream.range(0, _parameters.size())
+            .filter(i -> _parameters.get(i).getParamType() == Parameter.ParamType.CALLBACK)
+            .findFirst()
+            .orElse(-1);
+        // The actual param is wrapped inside the Callback as a generic param, so extract the embedded generic type.
+        return getFirstActualType((ParameterizedType) _method.getGenericParameterTypes()[callBackParamIndex]);
+      case TASK:
+      case PROMISE:
+        // Task and Promise return types wrap the actual type inside a task or promise, so extract the embedded
+        // generic type.
+        return getFirstActualType((ParameterizedType) _method.getGenericReturnType());
+      default:
+        return _method.getReturnType();
+    }
+  }
+
+  /**
+   * @return The first actual type from a parameterized type as a class.
+   */
+  private Class<?> getFirstActualType(ParameterizedType parameterizedType) {
+    Type unwrappedType = parameterizedType.getActualTypeArguments()[0];
+    // Now there are 2 cases. The generic type may represent a parameterized type itself, in which case we need
+    // to extract its raw type because we don't care about its generic type. Else we can just cast it to a class
+    // and return it.
+    if (unwrappedType instanceof ParameterizedType) {
+      return (Class<?>) ((ParameterizedType) unwrappedType).getRawType();
+    }
+    return (Class<?>) unwrappedType;
+  }
+
+  /**
+   * @return Resource Java method type.
+   */
+  public JavaMethodType getJavaMethodType()
```

Comment:
ResourceMethodDescriptor has _interfaceType can be directly mapped to your JavaMethodType.